### ISSUE TITLE
Model accuracy error

### ISSUE THREAD

**Issue by someclick** [facebookresearch/videoseal#33]
Hello, when testing model accuracy error with an image. I got bits-error too high in some models. And in my trained model, it got bits-error ~50% too. How to resolve it?
this is code for testing 

```
from PIL import Image as Image
import torch, torchvision
import numpy as np
from videoseal.evals.full import setup_model_from_checkpoint
to_tensor = torchvision.transforms.ToTensor()

def test(checkpoint, size=256):
    device = "cpu"
    model = setup_model_from_checkpoint(checkpoint)
    model = model.eval().to(device)
    imgs = Image.open("image.jpg", "r").convert("RGB")
    img = to_tensor(imgs).unsqueeze(0).float().to(device)
    if size > 0:
        img = img[:,:,:size,:size]
    msg = torch.randint(0, 2, (1, model.embedder.msg_processor.nbits)).float().to(device)
    embeds = model.embed(img, msg, is_video = False)
    embed = embeds['imgs_w']
    detect = (model.detect(embed, is_video = False)["preds"][:, 1:].numpy()).flatten()
    detect = (detect > 0.0).astype(np.int32)
    print(f"{checkpoint:20s} bits error: {np.sum(np.abs(detect - msg.numpy()))*100/len(detect):.2f}%")

test("videoseal")
test("y_256b_img.pth")
test("y_128b_img_full.pth")
test("y_256b_img_full.pth")
```
and the result in log
```
videoseal            bits error: 1.95%
y_256b_img.pth       bits error: 2.73%
y_128b_img_full.pth  bits error: 53.12%
y_256b_img_full.pth  bits error: 51.95%
```

**Comment by pierrefdz** [comment]
Hi!
Thanks for flagging these

- For the mismatch between `videoseal` and  `y_256b_img,pth` which are actually the same models, this comes from the fact that the `scaling_w` parameter is updated to 0.2 in the model card but not in the checkpoint. 
You can either change the `scaling_w` at inference with `model.blender.scaling_w = 0.2` or create a new checkpoint with the update parameter:
```py
checkpoint = torch.load("ckpts/y_256b_img.pth", map_location='cpu', weights_only=True)
checkpoint["args"]["scaling_w"] = 0.2
checkpoint = {
    "args": checkpoint["args"],
    "model": checkpoint["model"],
}
torch.save(checkpoint, "ckpts/y_256b_img_solved.pth")
```

- For the mismatch between `y_256b_img.pth` and `y_256b_img_full.pth`, this comes from an error in the way I saved the models. You should have seen a warning about the weights not being properly loaded.
We saved the checkpoint as {"args": ..., "model": {"model":..., "discriminator":..., ...}} and not as {"args": ..., "model": ..., "discriminator":...}. 

If you want to do it on your side, you can solve it with:
```
checkpoint = torch.load("ckpts/y_128b_img_full.pth", map_location='cpu', weights_only=True)
checkpoint = {
    "args": checkpoint["args"],
    **checkpoint["model"]
}
torch.save(checkpoint, "ckpts/y_256b_img_solved.pth")
checkpoint.keys()
```

I'll update the hosted weights on our side early next week. 

- For the errors on `test("videoseal")` I am not able to reproduce with the image of the repo (`assets/imgs/1.jpg`). Running it on my side gives `videoseal            bits error: 0.00%`.
My guess is that it comes from your image. Is it a particular image? For instance, this can maybe happen on very flat / smooth images because of the attenuation that we apply that focuses on textures?

**Comment by someclick** [comment]
Hello, thank you for your quick response.

Regarding the checkpoint issue, I also figured it out by displaying the keys of the checkpoint.

For the test("videoseal"), the image I used was taken from the internet. The small error margin depends on the randomly generated message. 

As for the training issue, I used a sample dataset of around 10,000 images (a [subset ](https://ai.meta.com/datasets/segment-anything-downloads/)of the SA dataset). After running about 40 epochs, the bit accuracy is still around 50% with no improvement. Do you have any recommendations regarding the image size and quantity that should be used?

```
torchrun --nproc_per_node=1 train.py --batch_size 8 --batch_size_eval 8 \
    --video_dataset none --image_dataset sa --workers 12 \
    --extractor_model convnext_tiny --embedder_model unet_small2_yuv_quant --hidden_size_multiplier 1 --nbits 32 \
    --scaling_w_schedule Cosine,scaling_min=0.2,start_epoch=200,epochs=400 --scaling_w 1.0 --scaling_i 1.0 --attenuation jnd_1_1 \
    --epochs 601 --iter_per_epoch 1000 --scheduler CosineLRScheduler,lr_min=1e-6,t_initial=601,warmup_lr_init=1e-8,warmup_t=20 --optimizer AdamW,lr=5e-4 \
    --lambda_dec 1.0 --lambda_d 0.1 --lambda_i 0.1 --perceptual_loss yuv  --num_augs 2 --augmentation_config configs/all_augs.yaml --disc_in_channels 1 --disc_start 50 
```

> INFO:root:Train - Epoch: [39/601] - Modality: img  [ 920/1249]  eta: 0:01:25  total_loss: 0.6932 (0.6932)  loss_percep: 0.0000 (0.0000)  loss_disc: 0.7492 (0.7526)  loss_decode: 0.6932 (0.6932)  scale_percep: 0.1000 (0.1000)  scale_disc: 0.0000 (0.0000)  scale_decode: 1.0000 (1.0000)  psnr: 91.8535 (91.1710)  ssim: 1.0000 (1.0000)  lr: 0.0005 (0.0005)  bit_acc: 0.4805 (0.5000)  time: 0.247817  data: 0.000267  max mem: 4052

**Comment by someclick** [comment]
Hello Pierre, thanks for your help
I’ve updated the config and retrained the model — it seems like everything is working now. The accuracy is increasing slowly, which might be related to the number of images used for training.

I’d appreciate your feedback on the following points:
- Recommended number/resolution of images to use
- Minimum/maximum number of epochs (early stopping can be applied)
- Minimum bit accuracy that should be achieved

```
torchrun --nproc_per_node=1 train.py --batch_size 8 --batch_size_eval 8 \
    --video_dataset none --image_dataset sa --workers 12 \
    --extractor_model convnext_tiny --embedder_model unet_small2_yuv_quant \
    --hidden_size_multiplier 1 --nbits 128 \
    --scaling_w_schedule Cosine,scaling_min=0.2,start_epoch=200,epochs=400 \
    --scaling_w 1.0 --scaling_i 1.0 --attenuation jnd_1_1 \
    --epochs 601 --iter_per_epoch 1000 \
    --scheduler CosineLRScheduler,lr_min=1e-6,t_initial=601,warmup_lr_init=1e-8,warmup_t=20 \
    --optimizer AdamW,lr=1e-4 \
    --lambda_dec 1.0 --lambda_d 0.1 --lambda_i 0 --perceptual_loss yuv \
    --num_augs 1 --augmentation_config configs/all_augs.yaml \
    --disc_in_channels 1 --disc_start 50 

```

![Image](https://github.com/user-attachments/assets/96b42f3a-f7ba-4f25-b919-e21730e8207e)

**Comment by pierrefdz** [comment]
Hi!

Now that the training is working you can:
- try to increase the LR a little bit (until it breaks the optim). I think you can go up to 1e-3.
- do more steps during each epoch (so put more images in your folder)
- do more epochs (but make sure to change the schedulers accordingly)

In our case we train on 16GPUs with bsz of 16 per gpu  -- if I remember correctly -- so the training will be hard to reproduce perfectly. however you can try to decrease the number of bits or the augmentations if you want to reach 100% bit accuracy during the training.

**Important note**

This argument is particularly important`--scaling_w_schedule Cosine,scaling_min=0.2,start_epoch=200,epochs=400`. What it does here is that it schedules the scaling_w during training. It starts at `args.scaling_w` and plateaus until `start_epoch`. At `start_epoch=200` it starts decreasing to `scaling_min=0.2` and it does so in `epochs=400` epochs.
So the model should have converged to 100% bit acc around 300 epochs or so, before the `scaling_w` starts decreasing a lot.

If you change the number of epochs, make sure to change this parameters. What you can also do is train with a high `scaling_w`, and when you reach 100% training bit accuracy, you resume from this checkpoint and do the `scaling_w` schedule.

**Comment by llCurious** [comment]
@pierrefdz hi, i'm trying to train the 256-bit model with the following script

```
OMP_NUM_THREADS=40 torchrun --nproc_per_node=2 train.py --local_rank 0 \
    --video_dataset none --image_dataset myimages --workers 8 \
    --extractor_model convnext_tiny --embedder_model unet_small2_yuv_quant --hidden_size_multiplier 1 --nbits 256 \
    --scaling_w_schedule Cosine,scaling_min=0.2,start_epoch=200,epochs=400 --scaling_w 1.0 --scaling_i 1.0 --attenuation jnd_1_1 \
    --epochs 601 --iter_per_epoch 1000 --scheduler CosineLRScheduler,lr_min=1e-6,t_initial=601,warmup_lr_init=1e-8,warmup_t=20 --optimizer AdamW,lr=1e-4 \
    --lambda_dec 1.0 --lambda_d 0.1 --lambda_i 0 --perceptual_loss yuv  --num_augs 2 --augmentation_config configs/all_augs.yaml --disc_in_channels 1 --disc_start 50 \
    --output_dir output
```

Specifically, `myimages` refers to a sample dataset of around 10,000 images (a [subset ](https://ai.meta.com/datasets/segment-anything-downloads/)of the SA dataset). However, the convergence is extremely slow 

<img width="2810" height="194" alt="Image" src="https://github.com/user-attachments/assets/f86a31fa-dccd-4974-8d5c-c0e1febaa103" />

Do you have any ideas or suggestions? I have tried (as the discussion above, the mere difference is 128 v.s. 256 bit training)
1. using a small LR `1e-4`; 
2. set `--lambda_i 0`;

**Comment by llCurious** [comment]
hi @pierrefdz . Do you mind providing the training script for the description in the paper. Thx a lot in advance!

`For the last 300 epochs, we also use the SA-V dataset, with 200 steps per epoch. We only forward one 32-frame clip per GPU, randomly chosen at every step. The learning rate is linearly increased from 10−7 to 10−6 over the first 10 epochs, and then follows a cosine schedule down to 10−8 until the last epoch. At epoch 250, we freeze the embedder and only optimize the extractor.`
